packages/react-pose/src/components/Transition/children.ts

```diff
@@ -1,58 +1,69 @@
 import * as React from 'react';
 import { ReactElement } from 'react';
-import { CurrentPose } from '../PoseElement/types';
-import { MergeChildrenProps, Props } from './types';
+import { Props, State } from './types';
+import { invariant } from 'hey-listen';
 const { Children, cloneElement } = React;
 
-const getKey = (child: ReactElement<any>): string => child.key as string;
-
-const filterChildProps = ({ children, _pose, onPoseComplete, popFromFlow, ...props }: Props) => props;
+const filterChildProps = ({
+  children,
+  _pose,
+  onPoseComplete,
+  popFromFlow,
+  ...props
+}: Props) => props;
+
+const getKey = (child: ReactElement<any>): string => {
+  invariant(
+    child && child.key !== null,
+    'Every child of Transition must be given a unique key'
+  );
 
-const animateChildrenList = (
-  incomingChildren: Array<ReactElement<any>>,
-  pose: CurrentPose,
-  initialPose?: CurrentPose
-) => {
-  const children: Array<ReactElement<any>> = [];
+  const childKey =
+    typeof child.key === 'number' ? child.key.toString() : child.key;
 
-  Children.forEach(incomingChildren, (child: ReactElement<any>) =>
-    children.push(cloneElement(child, { pose, initialPose }))
-  );
-  return children;
+  return childKey.replace('.$', '');
 };
 
-const mergeChildren = ({
-  incomingChildren,
-  displayedChildren,
-  isLeaving,
-  removeFromTree,
-  groupProps
-}: MergeChildrenProps) => {
-  const {
-    children: groupChildren,
+const handleTransition = (
+  {
+    children: targetChildren,
     preEnterPose,
     enterPose,
     exitPose,
-    flipMove,
     animateOnMount,
+    enterAfterExit,
+    flipMove,
+    onRest,
     ...propsForChild
-  } = groupProps;
+  }: Props,
+  {
+    children: displayedChildren,
+    finishedLeaving,
+    scheduleChildRemoval,
+    hasMounted
+  }: State
+) => {
+  targetChildren = makeChildList(targetChildren);
+
   const children: Array<ReactElement<any>> = [];
 
   const prevKeys = displayedChildren.map(getKey);
-  const nextKeys = incomingChildren.map(getKey);
+  const nextKeys = targetChildren.map(getKey);
 
   const entering = new Set(
-    nextKeys.filter(key => isLeaving.has(key) || prevKeys.indexOf(key) === -1)
+    nextKeys.filter(
+      key => finishedLeaving.hasOwnProperty(key) || prevKeys.indexOf(key) === -1
+    )
   );
-  entering.forEach(key => isLeaving.delete(key));
+  entering.forEach(key => delete finishedLeaving[key]);
```

Comment:
im curious about the design here - `entering` is still a set, but `finishedLeaving` became a POJO, whats the reasoning behind this change?

Reply:
`finishedLeaving` could equally be a `Map` I guess, maybe it will be, but the difference is `finishedLeaving` holds boolean data whereas `entering` doesn't.